### PR TITLE
ConnectionStateListener change

### PR DESCRIPTION
**Pull request by xinlian12**
This PR is mostly a duplicate of ConnectionStateListener change incuded in PR https://github.com/Azure/azure-sdk-for-java/pull/14697 with few differences.

**Implementation**:
Each `RntbdServiceEndpoint` has a `RntbdConnectionStateListener` to keep tracking the partitionKeyRangeIdentity sets. New partitionKeyRangeIdentity will be added when `RntbdServiceEndpoint.request()` called.

When detecting a server is going down, we will remove all the effected `PartitionKeyRangeIdentity` from gateway address cache. Currently, only ClosedChannelException will trigger `onConnectionEvent` since this is more sure as a signal the server is going down. 

**Workflow**:
Three major workflows could be triggered.
1. Server starts a graceful connection closure. (FIN)
    Normal netty close channle flow will be triggered. (closeFuture, inactive, unregister). `RntbdRequestManager` will complete all pending requests for the channel with `ClosedChannelException`. 

2. Server ungracefully close a connection (RST).
    This will trigger netty exceptionCaught() flow. 'RntbdRequestManager.exceptionCaught()` will be called, and complete all pending requests for the channel with IOException. Channel will be closed.

3. When the server down, client trying to start a new connection:
   Eventually get ConnectTimeOutException

**Test**: 
Test33: 
Did TCP packets capture for package upgrade and manually retart VM. Based on the tcp traces, the first and third workflow will be triggered.

**Test Results from benchmark run:**
Pending

**Test Results from CTL run:**
Pending

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RequestTimeline.java [hunk 1]
```diff
@@ -157,13 +157,11 @@ public String toString() {
         @JsonIgnore
         private final Duration duration;
 
-        @JsonSerialize(using = ToStringSerializer.class)
-        private final long durationInMicroSec;
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RequestTimeline.java:161]: Why are we removing this ?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdConnectionEvent.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+//------------------------------------------------------------
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdConnectionEvent.java:1]: This header looks different from our other headers.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RequestTimeline.java (hunk 1 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RequestTimeline.java:161]: _"It is reported, just not stored as a value"_ (copied from the original PR).

This will not change the rntbd timeline diagnostics. After this change: 
"serializationDiagnosticsContext":{"serializationDiagnosticsList":[{"serializationType":"ITEM_SERIALIZATION","startTimeUTC":"22 Sep 2020 17:40:59.239","endTimeUTC":"22 Sep 2020 17:40:59.239","durationInMicroSec":0},{"serializationType":"PARTITION_KEY_FETCH_SERIALIZATION","startTimeUTC":"22 Sep 2020 17:40:59.317","endTimeUTC":"22 Sep 2020 17:40:59.323","durationInMicroSec":5998}]},"gatewayStatistics":null,"systemInformation":{"usedMemory":"60586 KB","availableMemory":"8309590 KB","systemCpuLoad":"(2020-09-22T17:40:58.729927700Z 20.1%)"}}

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdConnectionEvent.java (hunk 2 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdConnectionEvent.java]: Thanks, changed to be same as other files.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java [hunk 3]
```diff
@@ -32,6 +33,7 @@
      * Constructor
      */
     public DirectConnectionConfig() {
+        this.connectionEndpointRediscoveryEnabled = true;
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java]: added DEFAULT_CONNECTION_ENDPOINT_REDISCOVERY_ENABLED, and mentioned in the getter and setter method.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java [hunk 4]
```diff
@@ -123,18 +121,18 @@
         new ClosedChannelException(), RntbdClientChannelPool.class, "acquire");
 
     private static final IllegalStateException POOL_CLOSED_ON_ACQUIRE = ThrowableUtil.unknownStackTrace(
-        new StacklessIllegalStateException("service endpoint was closed while acquiring a channel"),
+        new ChannelAcquisitionException("service endpoint was closed while acquiring a channel"),
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java:124]: do we translate `ChannelAcquisitionException` to GoneException?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdConnectionStateListener.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+                // * a request timed out in pending acquisition queue
+                // * a request failed fast in admission control layer due to high load
+                // * channel connect timed out
+                //
+                // Currently, only ClosedChannelException will raise onConnectionEvent since it is more sure of a signal the server is going down.
+
+                if (cause instanceof IOException) {
+                    final Class<?> type = cause.getClass();
+
+                    if (type == ClosedChannelException.class) {
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdConnectionStateListener.java:71]: any reason this is preferred over `cause instance of ClosedChannelException` ?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequest.java [hunk 6]
```diff
@@ -54,7 +55,7 @@ public static RntbdRequest decode(final ByteBuf in) {
             throw new IllegalStateException(reason);
         }
 
-        final int start = in.readerIndex();
+        final int start = in.writerIndex();
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequest.java:58]: why? is this related to this PR? or an unrelated bug?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java [hunk 7]
```diff
@@ -506,7 +506,7 @@ public void write(final ChannelHandlerContext context, final Object message, fin
 
         if (message == RntbdHealthCheckRequest.MESSAGE) {
 
-            context.write(RntbdHealthCheckRequest.MESSAGE, promise).addListener(completed -> {
+            context.writeAndFlush(RntbdHealthCheckRequest.MESSAGE, promise).addListener(completed -> {
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java:509]: why `flush` is required here? I don't think this is in the scope of the connection listener. am I right?

please undo this if unrelated to this PR.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdConnectionStateListener.java [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+import static com.azure.cosmos.implementation.guava25.base.Preconditions.checkNotNull;
+
+public class RntbdConnectionStateListener {
+    // region Fields
+
+    private static final Logger logger = LoggerFactory.getLogger(RntbdConnectionStateListener.class);
+
+    private final IAddressResolver addressResolver;
+    private final RntbdEndpoint endpoint;
+    private final Set<PartitionKeyRangeIdentity> partitionAddressCache;
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdConnectionStateListener.java:29]: `RntbdConnectionStateListener:partitionAddressCache`  is a local cache and can be outdated. as a result of that `RntbdConnectionStateListener` may remove addresses of unrelated pkranges causing unnecessary refresh.

Consider this scenario:
1. PKR1, PKR2, PKR3 are hosted on the same physical node.
2. request1 with PKR1, request2 with PKR2, and request3 with PKR3 are sent and as a result RntbdConnectionStateListener: partitionAddressCache is populated.
3. due to  partition movement PKR1 moves out of the physical node, but PKR2 and PKR3 stay on the node.
4. now the node hosting PKR2 and PKR3 shuts down.
5. RntbdConnectionStateListener thinks PKR1 is still on this physical node, and hence will correctly remove addresses of PKR1, PKR2, and PKR3. But PKR3 addresses shouldn't have been removed!!.

**Review by moderakh — CHANGES_REQUESTED** [review]
`RntbdConnectionStateListener::partitionAddressCache`  is a local cache and can be stale. as a result of that `RntbdConnectionStateListener` may remove addresses of unrelated pkranges causing unnecessary refresh.

Consider this scenario:
1. PKR1, PKR2, PKR3 are hosted on the same physical node.
2. request1 with PKR1, request2 with PKR2, and request3 with PKR3 are sent and as a result `RntbdConnectionStateListener:partitionAddressCache` is populated with PKR1, PKR2, and PKR3.
3. due to  partition movement PKR1 moves out of the physical node, but PKR2 and PKR3 stay on the node.
4. now the node hosting PKR2 and PKR3 shuts down.
5. RntbdConnectionStateListener thinks PKR1 is still on this physical node, and hence will remove addresses of PKR1, PKR2, and PKR3. But PKR1 addresses shouldn't have been removed!!.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java [hunk 9]
```diff
@@ -40,6 +43,47 @@ public DirectConnectionConfig() {
...
     }
 
+    /**
+     * Gets a value indicating whether Direct TCP connection endpoint rediscovery is enabled.
+     * <p>
+     * The connection endpoint rediscovery feature is designed to reduce and spread-out latency spikes that are likely
+     * to occur:
+     * <ul>
+     * <li>During rolling upgrades of a Cosmos instance or
+     * <li>When a backend node is being decommissioned or restarted (e.g., to restart or remove an unhealthy replica.)
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java]: Thanks Mo, updated for both getter and setter method.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java [hunk 10]
```diff
@@ -40,6 +43,47 @@ public DirectConnectionConfig() {
...
+    }
+
+    /**
+     * Sets a value indicating whether Direct TCP connection endpoint rediscovery should be enabled.
+     * <p>
+     * The connection endpoint rediscovery feature is designed to reduce and spread-out latency spikes that are likely
+     * to occur:
+     * <ul>
+     * <li>During rolling upgrades of a Cosmos instance or
+     * <li>When a backend node is being decommissioned or restarted (e.g., to restart or remove an unhealthy replica.)
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java]: Updated.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/AddressResolver.java [hunk 11]
```diff
@@ -62,6 +64,11 @@ public void initializeCaches(
         this.collectionRoutingMapCache = collectionRoutingMapCache;
     }
 
+    @Override
+    public void remove(RxDocumentServiceRequest request, Set<PartitionKeyRangeIdentity> partitionKeyRangeIdentitySet) {
+        throw new NotImplementedException("remove() is not supported in AddressResolver");
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/AddressResolver.java:69]: Rntbd use GlobalAddressResolver, so by theory, we should never hit it.

But since I add the remove() at the interface, so has to implement it.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/RntbdTransportClient.java [hunk 12]
```diff
@@ -300,6 +369,9 @@ private void throwIfClosed() {
         @JsonProperty()
         private final Duration connectionAcquisitionTimeout;
 
+        @JsonProperty()
+        private final boolean connectionEndpointRediscoveryEnabled;
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/RntbdTransportClient.java:335]: Added.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/RntbdTransportClient.java [hunk 13]
```diff
@@ -525,7 +599,6 @@ public String toString() {
          *   "maxRequestsPerChannel": 30,
          *   "maxConcurrentRequestsPerEndpointOverride": 500,
          *   "receiveHangDetectionTime": "PT1M5S",
-         *   "requestExpiryInterval": "PT5S",
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/RntbdTransportClient.java]: My understanding is that this config is not used anywhere. 
Also moved this change to PR https://github.com/Azure/azure-sdk-for-java/pull/16197 to keep this PR clean

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java (hunk 4 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java]: Yes, it will be translated into GoneException.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequest.java (hunk 6 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequest.java]: As mentioned in the original PR https://github.com/Azure/azure-sdk-for-java/pull/14697, we have to use writerIndex. 
Also moved this change to PR https://github.com/Azure/azure-sdk-for-java/pull/16197

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdConnectionStateListener.java (hunk 8 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdConnectionStateListener.java:29]: Yea, based on the current implementation, it could happen. 
Will think more and check how .net implement and then update here.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RequestTimeline.java (hunk 1 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RequestTimeline.java]: This change has moved to PR https://github.com/Azure/azure-sdk-for-java/pull/16197 to keep this PR clean.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/RntbdTransportClient.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/RntbdTransportClient.java:335]: cere -> cer to be consistent with other true/false formatting names.

[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/RntbdTransportClient.java:335]: updated.

'e' is for enabled, so we should ignore it?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdConnectionStateListener.java (hunk 5 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdConnectionStateListener.java]: changed to instance of ClosedChannelException

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdConnectionStateListener.java (hunk 8 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdConnectionStateListener.java:29]: let's track this as an optimization for later. (out of the scope of this PR)

**Review by moderakh — APPROVED** [review]
LGTM. (please make sure we run CTL)

Please track the following as an improvement outside of the scope of this PR:

**`RntbdConnectionStateListener::partitionAddressCache`  is a local cache and can be stale. as a result of that `RntbdConnectionStateListener` may remove addresses of unrelated pkranges causing unnecessary address refresh.**

Consider this scenario:
1. PKR1, PKR2, PKR3 are hosted on the same physical node.
2. request1 with PKR1, request2 with PKR2, and request3 with PKR3 are sent and as a result `RntbdConnectionStateListener:partitionAddressCache` is populated with PKR1, PKR2, and PKR3.
3. due to  partition movement PKR1 moves out of the physical node, but PKR2 and PKR3 stay on the node.
4. now the node hosting PKR2 and PKR3 shuts down.
5. RntbdConnectionStateListener thinks PKR1 is still on this physical node, and hence will remove addresses of PKR1, PKR2, and PKR3. But PKR1 addresses shouldn't have been removed!!.
6. this will result in unnecessary address refresh for PKR1.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java (hunk 7 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java]: Has removed this change to PR https://github.com/Azure/azure-sdk-for-java/pull/16197

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdConnectionStateListener.java (hunk 8 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdConnectionStateListener.java:29]: Created github working item to track this issue:
1. https://github.com/Azure/azure-sdk-for-java/issues/16245
2. https://github.com/Azure/azure-sdk-for-java/issues/16246

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java [hunk 14]
```diff
@@ -40,6 +43,37 @@ public DirectConnectionConfig() {
...
+    /**
+     * Gets a value indicating whether Direct TCP connection endpoint rediscovery is enabled.
+     * <p>
+     * The connection endpoint rediscovery feature is designed to reduce and spread-out latency spikes that may occur during maintenance operations.
+     *
+     * By default, connection endpoint rediscovery is disabled.
+     *
+     * @return {@code true} if Direct TCP connection endpoint rediscovery is enabled; {@code false} otherwise.
+     */
+    public boolean isConnectionEndpointRediscoveryEnabled() {
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:55]: please add Beta annotation.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java [hunk 15]
```diff
@@ -40,6 +43,37 @@ public DirectConnectionConfig() {
...
+     * The connection endpoint rediscovery feature is designed to reduce and spread-out latency spikes that may occur during maintenance operations.
+     *
+     * By default, connection endpoint rediscovery is disabled.
+     *
+     * @param connectionEndpointRediscoveryEnabled {@code true} if connection endpoint rediscovery is enabled; {@code
+     *                                             false} otherwise.
+     *
+     * @return the {@linkplain DirectConnectionConfig}.
+     */
+    public DirectConnectionConfig setConnectionEndpointRediscoveryEnabled(boolean connectionEndpointRediscoveryEnabled) {
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:71]: please add Beta annotation.

**Review by moderakh — COMMENTED** [review]
The PR looks good code-wise.

There are a few improvement which Annie is tracking outside of the scope of this PR.

for This PR however we are waiting for
1. CTL result
2. perf testing (there was some discussion that on dotNet side enabling this feature resulted in perf impact) we need to measure if that's the case in Java as well or not.

**Comment by xinlian12** [comment]
Test Result:

1. Adobe Tests (default config):
image
image
image

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java (hunk 15 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:74]: Added. Thanks

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java (hunk 14 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:57]: Added, thanks
